Refactor transaction handling

arkgil opened the pull request:
Unfortunately handling transactions asynchronously required pretty big changes to how we processed STUN messages.

Changes:
* transactions are handled asynchronously now
* permissions are now added to worker's state when we send them and acknowledged when we receive asynchronous success response
* permissions, relay data and credentials now have their own structs
* spli huge `Protocol` module to couple of modules each handling one STUN method

Commit:
Keep credentials data in separate struct

Commit:
Keep relay related data in sepatate struct

Commit:
Handle server responses asynchronously

Also split Protocol modules into modules
handling different methods.

Commit:
Add tests for new protocol modules

Commit:
Fix dialyzer warnings on unknown types

rslota commented on lib/jerboa/client.ex:
```diff
@@ -109,15 +109,24 @@ defmodule Jerboa.Client do
   @spec refresh(t) :: :ok | {:error, error}
   def refresh(client) do
-    GenServer.call(client, :refresh)
+    call = fn -> GenServer.call(client, :refresh) end
```
How about `maybe_retry(request/1, [:refresh])`? This block is used several times, so splitting it into two functions feels like a good idea. 

```elixir
defp request(req), do: GenServer.call(client, req)
defp maybe_retry(fun, req) do
  case fun(req) do
    {:error, :stale_nonce} -> fun(req)
    {:error, :unauthorized} -> fun(req)
    result -> result
  end
end

erszcz commented on lib/jerboa/client/credentials.ex:
```diff
@@ -0,0 +1,29 @@
+
+  defstruct [:username, :realm, :secret, :nonce]
+
+  @type t :: %__MODULE__{
```
I'd be keen on seeing this refactored in the light of this idea: https://www.youtube.com/watch?v=kZ1P8cHN3pY&feature=youtu.be&t=21m27s
@arkgil what do you think about trying to do so? I'm not sure how many state changes we do before all fields are  defined, so it might not be worth the effort in this particular case. I'm leaving it to your judgment.

arkgil replied:
This looks really interesting! I'll try to implement the approach from the video here.

rslota commented:
@arkgil Could you please rebase this branch onto master? I would really like to test the timeout issue I had with previous implementation, but current `fennec` requires newest `master` version of `jerboa`.

Commit:
Split Credentials struct in two separate structs

This way we are always sure when final set of credentials
is used and when we only have initial information.

rslota commented on lib/jerboa/client/worker.ex:
```diff
@@ -96,161 +103,220 @@ defmodule Jerboa.Client.Worker do
-    |> remove_permissions(peer_addrs)
-    |> add_permissions(peer_addrs)
+  @spec create_perm_response_handler :: Transaction.handler
+  def create_perm_response_handler do
```
Do we really need to return an unnamed function? Why can't we just use `&create_perm_response_handler/3`? This comment applies to all those handlers. If we would skip this `fn` wrapper, we get less nested function.

arkgil replied:
I like it because of a typespec, it shows clearly that we return transaction handler.

rslota commented on lib/jerboa/client/worker.ex:
```diff
@@ -96,161 +103,220 @@ defmodule Jerboa.Client.Worker do
+    %{state | transactions: [t | state.transactions]}
+  end
+
+  @spec find_transaction(state, id :: binary) :: Transaction.t | nil
```
If would be a bit simpler and potentially much faster to use map `transaction_id => Transaction.t`. What do you think about it?

arkgil replied:
Yep, that's true. I was worried about possibly same transaction IDs, but in the end we're not handling them anyway, so map should be better here.

rslota commented on lib/jerboa/client/worker.ex:
```diff
@@ -96,161 +103,220 @@ defmodule Jerboa.Client.Worker do
-    |> remove_permissions(peer_addrs)
-    |> add_permissions(peer_addrs)
+  @spec create_perm_response_handler :: Transaction.handler
+  def create_perm_response_handler do
```
This handler function is public while others are private. Is there a reason? If not, I suggest normalizing it.

erszcz commented on lib/jerboa/client/credentials.ex:
```diff
@@ -1,29 +1,56 @@
+      nonce: nonce
+    }
+  end
+  def finalize(%Final{} = creds, _, _), do: creds
```
Do we need this clause for convenience? After the split it seems superfluous and by eliminating it we could make the spec narrower, too.

erszcz commented:
No more comments from me - good stuff!

Commit:
Refactor some pieces into separate functions

Commit:
Keep transactions as a map instead of a list

arkgil commented:
Applied changes as you suggested (except one 🙈 ). Also rebased on master @rslota 🙂

erszcz commented:
Works when plugged in esl/fennec@a123fa2, too. Merging!
